### ISSUE TITLE
Potential race condition between full reconnect and publishTrack

### ISSUE THREAD

**Issue by davidzhao** [livekit/client-sdk-js#507]
### Describe the bug

During a full reconnection sequence, the last step we [republish](https://github.com/livekit/client-sdk-js/blob/main/src/room/Room.ts#L735) all of the tracks to the server.

At the same time, it's possible a user is explicitly trying to republish the same underlying Track. This could cause a race condition where the same track is published multiple times to the server, potentially causing further downstream issues.

We should keep track of pending publications in publishTrack and fulfill the promise once complete.

### Reproduction

N/A

### System Info

```shell
all systems
```

### Severity

serious, but I can work around it

**Comment by davidzhao** [comment]
That's a good idea.. that'll mean we'd have to block user-initiated publishing during reconnection.

**Comment by burzomir** [comment]
@davidzhao @lukasIO I believe I have a similar issue but my exact problem (I guess) is that tracks are republished before signal connection is reestablished and this results in them not being republished at all despite livekit-client thinking that they are.

Console logs:
<img width="655" alt="image" src="https://user-images.githubusercontent.com/13577810/221864675-be06ad86-8278-4cee-bcf6-1a1cf180866b.png">

LocalParticipant:
<img width="779" alt="image (1)" src="https://user-images.githubusercontent.com/13577810/221865243-62dd4871-fa4f-474c-a615-9e81adcaf824.png">

RemoteParticipant:
<img width="724" alt="image (2)" src="https://user-images.githubusercontent.com/13577810/221865399-f8b71554-2efc-4d56-bcde-471648322bcb.png">

**Comment by burzomir** [comment]
I forgot I can set log level to debug :/ Now I think my issue may be caused by `RTCEngine.ts:260 failed to remove track` error. Tough I'm not sure what could cause that.
```
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VC4AkgEeQPMpoR, visible=false {trackSid: 'TR_VC4AkgEeQPMpoR'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: true, quality: 0, width: 212, height: 136, …}
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VSbsZHJ3k47kpN, visible=false {trackSid: 'TR_VSbsZHJ3k47kpN'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: true, quality: 0, width: 263, height: 717, …}
RTCEngine.ts:317 primary PC state changed disconnected
RTCEngine.ts:340 secondary PC state changed disconnected
RTCEngine.ts:317 primary PC state changed failed
RTCEngine.ts:685 primary peerconnection disconnected
RTCEngine.ts:714 reconnecting in 0ms
RTCEngine.ts:845 resuming signal connection, attempt 0
Room.ts:1401 room event connectionStateChanged {event: 'connectionStateChanged', args: Array(1)}
App.elm:509 livekit event: ConnectionStateChanged Reconnecting
Room.ts:1401 room event reconnecting {event: 'reconnecting', args: Array(0)}
SignalClient.ts:620 clearing ping interval
SignalClient.ts:210 connecting to <redacted>
SignalClient.ts:620 clearing ping interval
RTCEngine.ts:340 secondary PC state changed failed
RTCEngine.ts:685 secondary peerconnection disconnected
RTCEngine.ts:714 reconnecting in 0ms
SignalClient.ts:620 clearing ping interval
SignalClient.ts:613 start ping interval
RTCEngine.ts:685 leave-reconnect disconnected
RTCEngine.ts:714 reconnecting in 0ms
PCTransport.ts:96 restarting ICE
PCTransport.ts:118 starting to negotiate
PCTransport.ts:200 setting munged local description
SignalClient.ts:351 sending offer RTCSessionDescription {type: 'offer', sdp: 'v=0\r\no=- 6861072250490297233 3 IN IP4 127.0.0.1\r\ns…le\r\na=sctp-port:5000\r\na=max-message-size:262144\r\n'}
RTCEngine.ts:891 waiting for peer connection to reconnect
RTCEngine.ts:685 reconnect disconnected
RTCEngine.ts:714 reconnecting in 1000ms
RTCEngine.ts:340 secondary PC state changed disconnected
RTCEngine.ts:802 reconnecting, attempt: 1
SignalClient.ts:620 clearing ping interval
SignalClient.ts:210 connecting to <redacted>
SignalClient.ts:270 ping config {timeout: 20, interval: 10}
SignalClient.ts:620 clearing ping interval
SignalClient.ts:613 start ping interval
RTCEngine.ts:891 waiting for peer connection to reconnect
RTCEngine.ts:382 received server offer {RTCSdpType: 'offer', signalingState: 'stable'}
PCTransport.ts:200 setting munged local description
SignalClient.ts:360 sending answer
RTCEngine.ts:317 primary PC state changed connecting
RTCEngine.ts:382 received server offer {RTCSdpType: 'offer', signalingState: 'stable'}
PCTransport.ts:200 setting munged local description
SignalClient.ts:360 sending answer
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VC4AkgEeQPMpoR, visible=true {trackSid: 'TR_VC4AkgEeQPMpoR'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: false, quality: 0, width: 212, height: 136, …}
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VSbsZHJ3k47kpN, visible=true {trackSid: 'TR_VSbsZHJ3k47kpN'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: false, quality: 0, width: 263, height: 717, …}
RTCEngine.ts:317 primary PC state changed connected
Room.ts:748 reconnected to server {region: ''}
Room.ts:1401 room event connectionStateChanged {event: 'connectionStateChanged', args: Array(1)}
App.elm:509 livekit event: ConnectionStateChanged Connected
Room.ts:1401 room event reconnected {event: 'reconnected', args: Array(0)}
App.elm:509 livekit event: Reconnected
Room.ts:1401 room event trackUnsubscribed {event: 'trackUnsubscribed', args: Array(3)}
App.elm:509 livekit event: TrackUnsubscribed (RemoteTrack (Track Camera False <internals>)) (RemoteParticipant { camera = Just (RemoteTrack (Track Camera False <internals>)), identity = "912", screenShare = Just (RemoteTrack (Track ScreenShare False <internals>)) } <internals>)
Room.ts:1401 room event trackSubscribed {event: 'trackSubscribed', args: Array(3)}
App.elm:509 livekit event: TrackSubscribed (RemoteTrack (Track Camera False <internals>)) (RemoteParticipant { camera = Just (RemoteTrack (Track Camera False <internals>)), identity = "912", screenShare = Just (RemoteTrack (Track ScreenShare False <internals>)) } <internals>)
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VC4AkgEeQPMpoR, visible=false {trackSid: 'TR_VC4AkgEeQPMpoR'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: true, quality: 0, width: 212, height: 136, …}
Room.ts:1401 room event trackUnsubscribed {event: 'trackUnsubscribed', args: Array(3)}
App.elm:509 livekit event: TrackUnsubscribed (RemoteTrack (Track ScreenShare False <internals>)) (RemoteParticipant { camera = Just (RemoteTrack (Track Camera False <internals>)), identity = "912", screenShare = Just (RemoteTrack (Track ScreenShare False <internals>)) } <internals>)
Room.ts:1401 room event trackSubscribed {event: 'trackSubscribed', args: Array(3)}
App.elm:509 livekit event: TrackSubscribed (RemoteTrack (Track ScreenShare False <internals>)) (RemoteParticipant { camera = Just (RemoteTrack (Track Camera False <internals>)), identity = "912", screenShare = Just (RemoteTrack (Track ScreenShare False <internals>)) } <internals>)
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VSbsZHJ3k47kpN, visible=false {trackSid: 'TR_VSbsZHJ3k47kpN'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: true, quality: 0, width: 263, height: 717, …}
LocalParticipant.ts:710 unpublishing track {track: LocalVideoTrack, method: 'unpublishTrack'}
RTCEngine.ts:260 failed to remove track {error: DOMException: Failed to execute 'removeTrack' on 'RTCPeerConnection': The sender was not created by…, method: 'removeTrack'}
removeTrack @ RTCEngine.ts:260
unpublishTrack @ LocalParticipant.ts:741
(anonymous) @ Room.ts:773
Room.handleRestarted @ Room.ts:771
emit @ events.js:159
restartConnection @ RTCEngine.ts:832
await in restartConnection (async)
attemptReconnect @ RTCEngine.ts:742
(anonymous) @ RTCEngine.ts:717
setTimeout (async)
CriticalTimers.setTimeout @ timers.ts:8
RTCEngine.handleDisconnect @ RTCEngine.ts:717
attemptReconnect @ RTCEngine.ts:770
await in attemptReconnect (async)
(anonymous) @ RTCEngine.ts:717
setTimeout (async)
CriticalTimers.setTimeout @ timers.ts:8
RTCEngine.handleDisconnect @ RTCEngine.ts:717
primaryPC.onconnectionstatechange @ RTCEngine.ts:334
LocalParticipant.ts:902 updating track subscription permissions {allParticipantsAllowed: true, participantTrackPermissions: Array(0)}
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VC4AkgEeQPMpoR, visible=true {trackSid: 'TR_VC4AkgEeQPMpoR'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: false, quality: 0, width: 212, height: 136, …}
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VSbsZHJ3k47kpN, visible=true {trackSid: 'TR_VSbsZHJ3k47kpN'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: false, quality: 0, width: 263, height: 717, …}
join.html:1 Uncaught (in promise) DOMException: The operation could not be performed and was aborted
join.html:1 Uncaught (in promise) DOMException: The operation could not be performed and was aborted
PCTransport.ts:118 starting to negotiate
RemoteTrackPublication.ts:267 adaptivestream video dimensions 212x136 {trackSid: 'TR_VC4AkgEeQPMpoR'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: false, quality: 0, width: 212, height: 136, …}
RemoteTrackPublication.ts:267 adaptivestream video dimensions 263x717 {trackSid: 'TR_VSbsZHJ3k47kpN'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: false, quality: 0, width: 263, height: 717, …}
PCTransport.ts:200 setting munged local description
SignalClient.ts:351 sending offer RTCSessionDescription {type: 'offer', sdp: 'v=0\r\no=- 5021020479784642386 2 IN IP4 127.0.0.1\r\ns…le\r\na=sctp-port:5000\r\na=max-message-size:262144\r\n'}
RTCEngine.ts:505 on data channel 1, _reliable
RTCEngine.ts:505 on data channel 3, _lossy
RTCEngine.ts:357 received server answer {RTCSdpType: 'answer', signalingState: 'have-local-offer'}
Room.ts:1401 room event localTrackUnpublished {event: 'localTrackUnpublished', args: Array(2)}
App.elm:509 livekit event: LocalTrackUnpublished (Just (LocalTrack (Track Camera False <internals>)))
RTCEngine.ts:340 secondary PC state changed connecting
RTCEngine.ts:340 secondary PC state changed connected
RTCEngine.ts:340 secondary PC state changed failed
RTCEngine.ts:685 secondary peerconnection disconnected
RTCEngine.ts:714 reconnecting in 0ms
RTCEngine.ts:845 resuming signal connection, attempt 0
Room.ts:1401 room event connectionStateChanged {event: 'connectionStateChanged', args: Array(1)}
App.elm:509 livekit event: ConnectionStateChanged Reconnecting
Room.ts:1401 room event reconnecting {event: 'reconnecting', args: Array(0)}
SignalClient.ts:620 clearing ping interval
SignalClient.ts:210 connecting to <redacted>
SignalClient.ts:620 clearing ping interval
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VC4AkgEeQPMpoR, visible=false {trackSid: 'TR_VC4AkgEeQPMpoR'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: true, quality: 0, width: 212, height: 136, …}
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VSbsZHJ3k47kpN, visible=false {trackSid: 'TR_VSbsZHJ3k47kpN'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: true, quality: 0, width: 263, height: 717, …}
SignalClient.ts:620 clearing ping interval
SignalClient.ts:613 start ping interval
PCTransport.ts:96 restarting ICE
PCTransport.ts:118 starting to negotiate
PCTransport.ts:200 setting munged local description
SignalClient.ts:351 sending offer RTCSessionDescription {type: 'offer', sdp: 'v=0\r\no=- 5021020479784642386 3 IN IP4 127.0.0.1\r\ns…le\r\na=sctp-port:5000\r\na=max-message-size:262144\r\n'}
RTCEngine.ts:891 waiting for peer connection to reconnect
RTCEngine.ts:382 received server offer {RTCSdpType: 'offer', signalingState: 'stable'}
Room.ts:1401 room event connectionQualityChanged {event: 'connectionQualityChanged', args: Array(2)}
RTCEngine.ts:357 received server answer {RTCSdpType: 'answer', signalingState: 'have-local-offer'}
PCTransport.ts:200 setting munged local description
SignalClient.ts:360 sending answer
Room.ts:1401 room event connectionStateChanged {event: 'connectionStateChanged', args: Array(1)}
App.elm:509 livekit event: ConnectionStateChanged Connected
Room.ts:1401 room event reconnected {event: 'reconnected', args: Array(0)}
App.elm:509 livekit event: Reconnected
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: true, quality: 0, width: 212, height: 136, …}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: true, quality: 0, width: 263, height: 717, …}
LocalParticipant.ts:902 updating track subscription permissions {allParticipantsAllowed: true, participantTrackPermissions: Array(0)}
Room.ts:1401 room event connectionQualityChanged {event: 'connectionQualityChanged', args: Array(2)}
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VC4AkgEeQPMpoR, visible=true {trackSid: 'TR_VC4AkgEeQPMpoR'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: false, quality: 0, width: 212, height: 136, …}
RemoteTrackPublication.ts:259 adaptivestream video visibility TR_VSbsZHJ3k47kpN, visible=true {trackSid: 'TR_VSbsZHJ3k47kpN'}
RemoteParticipant.ts:57 send update settings {trackSids: Array(1), disabled: false, quality: 0, width: 263, height: 717, …}
Room.ts:1401 room event connectionQualityChanged {event: 'connectionQualityChanged', args: Array(2)}
Room.ts:1401 room event connectionQualityChanged {event: 'connectionQualityChanged', args: Array(2)}
Room.ts:1401 room event connectionQualityChanged {event: 'connectionQualityChanged', args: Array(2)}
```

**Comment by davidzhao** [comment]
@burzomir your issue is likely related to #595, which has been released in 1.6.7. Do you want to give that version a try and see if that resolves it?

**Comment by burzomir** [comment]
Thanks @davidzhao, I already tried. It seems to solve the issue with tracks but with 1.6.7 the long initial connection time issue is back :/ So again users who previously experienced PeerConnection timeouts,  have to wait more than 1 minute to connect to a room. And with update to livekit-server 1.3.5 they PeerConnection timeout issue is fully back ;( Currently the stack that works for us but has these race conditions is js-client@1.6.3 and server@1.3.4.
